modules/4-graphql.livemd

```diff
@@ -2,89 +2,160 @@
+GraphQL is a query language used to interact with and retrieve data from an application's data sources. It's structure is desinged for flexible and precise queries that efficiently interact with complex, highly nested data sets.  Using GraphQL, information is retreived by stepping through data as if it were arranged as a group of connected nodes instead of a strictly hierarchical set up. Think more of a labyrinth than a tree.  GraphQL can be implemented as a component of an application's API and has two main security considerations:
 
-*TODO: Write Introduction*
+* Security concerns common to all APIs 
+* Security related to characteristics of the query language itself
+
+This module will highlight several security issues associated with GraphQL and recommendations for how to address. 
 
 ## Table of Contents
 
 * [Disabling Introspection](#disabling-introspection)
 * [Error Disclosure](#error-disclosure)
-* [Resource Exhaustion](#resource-disclosure)
-  * [Cost Theory](#cost-theory)
+* [Resource Exhaustion](#resource-exhaustion)
 
 ## Disabling Introspection
 
 ### Description
 
-*TODO: Write Description*
+Introspection queries are a way of enumerating a particular graphql implementation to discover details about the queries supported, data types available, and other information.  This includes mutation names, fields specific to an organization/dataset, query parameters, and types of objects in the data source,  all of which can help a user, including a malicious one, deduce and discover specifics about the data being stored.   If you are familiar with databases, this is similar to gathering info on the database schema that includes information about table names, fields, database, structure etc.  https://en.wikipedia.org/wiki/Database_schema
+
+This information can help a malicious actor in their information gathering/reconnaissnce efforts as they look for ways to attack your application and construct malicious queries and requests to compromise data.
 
 ### Prevention
 
-*TODO: Write Prevention*
+Best practice per OWASP is to limit access, including following least privilege, to introspection queries if it is not possible completely disable it. Please see references for more details.
 
-### <span style="color:blue;">Example</span> / <span style="color:red;">Quiz</span>
+### References
+https://owasp.org/www-project-web-security-testing-guide/latest/4-Web_Application_Security_Testing/12-API_Testing/01-Testing_GraphQL
 
-*TODO: Make Example or Quiz Question*
+https://cybervelia.com/?p=736
 
-```elixir
+### <span style="color:blue;">Example</span> / <span style="color:red;">Quiz</span>
 
+```
+Which of the OWASP API Security Top 10 2019 issues does disabling introspection queries address?
+a) API6:2019 Mass Assignment
+b) API10:2019 Insufficient Logging & Monitoring
+c) API3:2019 Excessive Data Exposure
+d) API4:2019 Lack of Resources & Rate Limiting
 ```
```

Comment:
Great first iteration of a quiz question! Couple of notes here:
- Quiz questions _ideally_ are answerable just from having read the preceding section alone - it shouldn't be necessary to go hunt through an external resource to complete the training. 
  - So either that means you could reframe the Quiz question with that in mind or introduce the specific topic you're asking about in the preceding section.
  - Basically the only prior knowledge we should expect the user to have coming into this training is basic Elixir programming knowledge
- The formatting of the quiz question should be in line with the standard outlined in 1-introduction.livemd 
  - Basically just that the question should be outside of the code block in bold and on the following line in italics instructions on how to complete the quiz question.
- The way this particular quiz question is laid out won't actually work in its current form - remember that all the quiz questions will need to be Elixir code blocks that when the code block is run in the livebook, the answer is returned.
  - For an example of how we've previously done multiple choice questions in this training, you can refer to how we did it in the [Quiz question for "Security Through Obscurity"](https://github.com/podium/elixir-secure-coding/blob/main/modules/7-anti-patterns.livemd#quiz).